### DISCUSSION TITLE
Transform hex/base64 string into Buffer

### DISCUSSION THREAD

**Discussion by pke** [colinhacks/zod#3312] [Q&A] [answered]
I have input that can look like this:

```ts
[
  { id: "gZZ", data: "ALAAAAAAAA==" }
  { id: "gZZ", data: "00b00000000000" }
  { id: "gZZ", data: [0,0xb0,0,0,0,0,0] }
]
```

And I would like to convert all `data` into `Buffer` with their respective `Buffer.from` variant:

``` 
Buffer.from("ALAAAAAAAA==", "base64")
Buffer.from("00b00000000000", "hex")
Buffer.from([0,0xb0,0,0,0,0,0])
```

I started by supporting `base64` in my schema:

```ts
const schema = z.array(
  z.object({
    id: z.string().min(3),
    data: z.string().transform((value) => Buffer.from(value, "base64")),
  })
)
```

How would I support the `hex` and `number array` variants? I think I would need some regex to test for hex string input (and length needs to be even) and construct the Buffer depending on that.

**Comment by JacobWeisenburger** [comment]
This doesn't really seem like a question for zod. It seems that you are asking how to detect base64 or hex strings without needing to specify the encoding. This is a great question for which I have no answer. Perhaps it is better suited to be asked somewhere else. Perhaps stack overflow might be more helpful.

**Reply by pke** [reply]
I think it is related, as I am asking about transforms in the zod schema. See my answer below.

**Comment by pke** [comment, marked as answer]
I came up with this:

```ts

const hexString = z
  .string()
  .regex(/[a-fA-F0-9]{2}/, "Not a hex string")
  .refine((s) => s.length % 2 === 0)
  .transform((s) => Buffer.from(s, "hex"))
  .refine((buffer) => buffer.length >= 5)

const base64String = z
  .string()
  .regex(/^[-A-Za-z0-9+/]*={0,3}$/)
  .transform((s) => Buffer.from(s, "base64"))
  .refine((buffer) => buffer.length >= 5)

export const schema = z.array(
  z.object({
    id: z.string().min(3),
    data: z.union([hexString, base64String]).or(
      z
        .array(z.number())
        .min(5)
        .transform((value) => Buffer.from(value))
    ),
  })
)
```
